Added option to run ADT calls over RFC

akreuzer opened the pull request:
Main effort is to split Connection in `sap/adt/core.py` into `ConnectionViaRFC` and `ConnectionViaHTTP`.

I don't have many test for this as this would require a Netweaver Server.

Commit:
Added option to run ADT calls over RFC

Main effort is to split Connection in sap/adt/core.py into ConnectionViaRFC and ConnectionViaHTTP.

jfilak commented on sap/adt/core.py:
```diff
@@ -1,22 +1,23 @@
-    TimedOutRequestError
-)
+from sap.rest.errors import (HTTPRequestError, UnexpectedResponseContent,
+                             UnauthorizedError, TimedOutRequestError)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done

jfilak commented on sap/adt/core.py:
```diff
@@ -161,10 +292,20 @@ def _handle_http_error(self, req, res):
+                  url,
+                  params=None,
+                  headers=None,
+                  body=None):
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -174,18 +315,31 @@ def _retrieve(self, session, method, url, params=None, headers=None, body=None):
 
-        mod_log().debug('Response %s %s:\n++++\n%s\n++++', method, url, res.text)
+        mod_log().debug('Response %s %s:\n++++\n%s\n++++', method, url,
+                        res.text)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -174,18 +315,31 @@ def _retrieve(self, session, method, url, params=None, headers=None, body=None):
+                              url,
+                              params=None,
+                              headers=None,
+                              body=None):
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -174,18 +315,31 @@ def _retrieve(self, session, method, url, params=None, headers=None, body=None):
+                                  url,
+                                  params=params,
+                                  headers=headers,
+                                  body=body)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -174,18 +315,31 @@ def _retrieve(self, session, method, url, params=None, headers=None, body=None):
 
-        if res.status_code == 403 and (not headers or headers.get('x-csrf-token', '') != 'Fetch'):
+        if res.status_code == 403 and (
+                not headers or headers.get('x-csrf-token', '') != 'Fetch'):
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -194,12 +348,16 @@ def _execute_with_session(self, session, method, url, params=None, headers=None,
                 self._build_adt_url('discovery'),
-                headers={'x-csrf-token': 'Fetch'}
-            )
+                headers={'x-csrf-token': 'Fetch'})
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -194,12 +348,16 @@ def _execute_with_session(self, session, method, url, params=None, headers=None,
+                                      url,
+                                      params=params,
+                                      headers=headers,
+                                      body=body)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -216,30 +374,38 @@ def _get_session(self):
             # requests.session.verify is either boolean or path to CA to use!
-            self._session.verify = os.environ.get('SAP_SSL_SERVER_CERT', self._session.verify)
+            self._session.verify = os.environ.get('SAP_SSL_SERVER_CERT',
+                                                  self._session.verify)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -216,30 +374,38 @@ def _get_session(self):
-                mod_log().info('Using custom SSL Server cert path: SAP_SSL_SERVER_CERT = %s', self._session.verify)
+                mod_log().info(
+                    'Using custom SSL Server cert path: SAP_SSL_SERVER_CERT = %s',
+                    self._session.verify)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -216,30 +374,38 @@ def _get_session(self):
-                mod_log().info('SSL Server cert will not be verified: SAP_SSL_VERIFY = no')
+                mod_log().info(
+                    'SSL Server cert will not be verified: SAP_SSL_VERIFY = no'
+                )
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -216,30 +374,38 @@ def _get_session(self):
             try:
-                response = self._execute_with_session(self._session, 'GET', url, headers=discovery_headers)
+                response = self._execute_with_session(
+                    self._session, 'GET', url, headers=discovery_headers)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -216,30 +374,38 @@ def _get_session(self):
+            response = self._execute_with_session(self._session,
+                                                  'GET',
+                                                  url,
+                                                  headers=discovery_headers)
```
Please, avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -216,30 +374,38 @@ def _get_session(self):
+                    'Using custom SSL Server cert path: SAP_SSL_SERVER_CERT = %s',
+                    self._session.verify)
             elif self._ssl_verify is False:
-                import urllib3
```
How is this change related to ADT over RFC?

akreuzer replied:
Not related. I reverted it.

jfilak commented on sap/adt/core.py:
```diff
@@ -1,22 +1,23 @@
 
 import requests
 from requests.auth import HTTPBasicAuth
+import urllib3
```
How is this change related to ADT over RFC?

akreuzer replied:
Reverted.

jfilak commented on requirements.txt:
```diff
@@ -1,3 +1,5 @@
 pyodata==1.7.0
 PyYAML==5.4.1
+
+urllib3~=1.26.7
```
Isn't urllib3 needed only if a user wants to use HTTP over SSL?

akreuzer replied:
True. I removed it must have slipped in with the rebasing.

jfilak commented on bin/sapcli:
```diff
@@ -76,8 +76,11 @@ def parse_command_line(argv):
+        "--rfc_group",  # name "group" is already in use
+        default=os.getenv("SAP_GROUP"),
+        help="Group (use if connecting via a message "
+        "server")
```
Can you keep the format of the core around, please?

akreuzer replied:
Done.

jfilak commented on bin/sapcli:
```diff
@@ -86,6 +89,16 @@ def parse_command_line(argv):
+        not in [None, 'n', 'no', 'false', 'off'],
+        help=
+        "Prefer doing rest call over SAP RFC client (to use SNC or if HTTP port is firewalled)"
+    )
```
Can you keep the format of the core around, please?

akreuzer replied:
Done.

jfilak commented on sap/adt/core.py:
```diff
@@ -87,13 +88,156 @@ def _get_collection_accepts(discovery_xml):
+                 password,
+                 port=None,
+                 ssl=True,
+                 verify=True):
```
Please avoid cosmetic changes.

akreuzer replied:
Done.

jfilak commented:
Good job! I was always wondering how that RFC HTTP tunneling works :) I am happy to merge once we get rid of those unnecessary style fixes and all tests and linters are satisfied.

Commit:
Address comments

Revert changes to formating
Revert to old requirements.txt and move urllib3 back

akreuzer commented on sap/adt/core.py:
```diff
@@ -161,10 +292,20 @@ def _handle_http_error(self, req, res):
+                               url,
+                               params=params,
+                               data=body,
+                               headers=headers)
```
Done.

akreuzer commented:
Thank you. I must have accidentally run `yapf` on the changes. 
I should have reverted all of the cosmetic changes and also the stuff with `urllib3`.

jfilak commented on bin/sapcli:
```diff
@@ -86,6 +86,11 @@ def parse_command_line(argv):
                             help="SAP Secure Login Client library (e.g. "
                                  "/Applications/Secure Login Client.app/Contents/MacOS/lib/libsapcrypto.dylib")
 
+    arg_parser.add_argument("--rest-over-rfc", action='store_true', dest="rest_over_rfc",
```
Yay, I didn't notice earlier. The generic name 'rest-over-rfc' suggest that all HTTP traffic will be routed via RFC but I do believe only ADT HTTP traffic can be tunneled via RFC. Am I right?

akreuzer replied:
That is a very interesting question. In theory, all of the HTTP requests can be tunneled because it directly connects to the HTTP dispatcher on Netweaver Server. However, I did not try with other types of requests and it is not implemented.

jfilak replied:
OK, so since the implementation does not route entire HTTP traffic via RFC, I suggest  to rename to "adt_rest_over_rfc".

akreuzer replied:
That is a good point. 
I think it is better to keep the name but show an error message. With this, the parameter would not have to change if this is also implemented for gtcs. This is ultimately your call.

jfilak commented:
Please update `doc/configuration.md` with all the recent congiration options you have added.

Commit:
Added new parameter to doc/configuration.md and fixed type in bin/sapcli

Commit:
Corrected code formatting

Commit:
Not implemented error for rfc-over-rest and odata/gtcs

akreuzer commented:
- Added new options to `doc/configuration.md`
- Fixed on typo in `bin/sapcli`.
- Fixed some issue with liniting.
- Added NotImplementedException to the case where `--rest-over-rfc` is used with odata/gtcs. (This is just to illustrated - if you don't like it I can revert this commit.)

Commit:
Fix CI errors

- Fix sapcli test (rename `group` to `rfc_group`
- Fix type hints for python 3.8 (Dict[...] instead of dict[..])

lucasborin commented:
If you do not mind, I would like to test it. However, pulling your forked version raised a syntax error: 
![image](https://user-images.githubusercontent.com/5233413/165618608-330adaff-fd0a-467f-8ca1-9b725f15f8ae.png)

Besides, I added these two entries to my cfg file:
```
export SAPCLI_HTTP_TIMEOUT=2000
export SAP_REST_OVER_RFC=yes
```

Any clue?

Commit:
Change type syntax to be compatible with python 3.8

akreuzer commented:
I think I know what this is:
```list[str]``` as a type definition is only allowed in Python 3.9+

For older version on has to use something like this
```
from typing import List
a: List[str]
```

@lucasborin I updated the code. 
Let me know if it works now.

lucasborin commented:
Running under Python 3.8.10, another issue now:

```
./sapcli atc run package -o checkstyle log_cmm_prc_formula
Traceback (most recent call last):
  File "./sapcli", line 23, in <module>
    sys.exit(sapcli.main(sys.argv))
  File "/home/lucasborin/sapcli/bin/sapcli", line 142, in main
    args = parse_command_line(argv)
  File "/home/lucasborin/sapcli/bin/sapcli", line 96, in parse_command_line
    for connection, cmd in sap.cli.get_commands():
  File "/home/lucasborin/sapcli/sap/cli/__init__.py", line 167, in get_commands
    return CommandsCache.commands()
  File "/home/lucasborin/sapcli/sap/cli/__init__.py", line 74, in commands
    import sap.cli.startrfc
  File "/home/lucasborin/sapcli/sap/cli/startrfc.py", line 8, in <module>
    import pyrfc
ModuleNotFoundError: No module named 'pyrfc'
```

lucasborin commented:
Unsure if I did something wrong, but my python is not able to find PyRFC anymore before pulling it. I'll try to redo my setup.
